File: examples/test_array/test_array.f90
PROGRAM TEST_ARRAY 
   include 'mpif.h'
   integer rank, size, ierror, tag, comm, status(MPI_STATUS_SIZE)
   integer restored_i = 0, restored = 0, inject_error = -1, i
   integer, dimension(1024*1024) :: actual_data
   CHARACTER(len=32) :: arg
   
   
   call MPI_INIT(ierror)
   comm = MPI_COMM_WORLD

   if (iargc() == 1) then
       call getarg(1, arg)
       read (arg,'(I10)') inject_error
       print *, 'Inject error at step ', inject_error, '.'
   endif

!$chk init comm(comm)
!$chk load(restored_i, actual_data)
   if (restored_i .ne. 0) then
       print *, 'Restored data from iteration ', restored_i , '. data[i] = ', actual_data(restored_i), '.'
       restored = 1
   end if

   do i = restored_i, 1024*1024
       actual_data(i) = i
!$chk store(i, actual_data) kind(0) id(i) level(mod(i,4)+1) if(mod(i,(1024*128)) .eq. 1)
       if (i .eq. inject_error .AND. restored .eq. 0) then
           print *, 'Injected error'
           call EXIT(-1)
       end if
   end do
!$chk shutdown

   do i = 1, 1024*1024
      if (actual_data(i) .ne. i) then
          call EXIT(-1)
      end if
   end do

   call MPI_FINALIZE(ierror)
END PROGRAM

File: examples/test_array/test_array.cpp
#include <iostream>
#include <cstdlib>
#include <tcl.h>
#include <cassert>

void error_handler(int err_code) 
{
    std::cout << "Error inside CheckpointLib. Error code: " << err_code << "." << std::endl;
    exit(-1);
}

int main(int argc, char **argv) 
{
    int err = MPI_Init(&argc, &argv);
    assert(err == MPI_SUCCESS);
    MPI_Comm comm = MPI_COMM_WORLD;

    int inject_error = -1;
    if(argc == 2) {
        inject_error = std::atoi(argv[1]);
	std::cout << "Inject error at step " << inject_error << "." << std::endl;
    }

    int i = 0;
    int data[1024*1024];
    bool restored = false;

    #pragma chk init comm(comm)
    {

        #pragma chk load(i, data)
        if(i != 0) {
            std::cout << "Restored data from iteration " << i << ". data[i] = " << data[i] << "." << std::endl;
            restored = true;
        }
        for(i; i < 1024*1024; i++) {
            data[i] = i;
            #pragma chk store(i, data) kind(CHK_FULL) id(i) level((i%4)+1) if((i%(1024*128))==1) handler(error_handler)
            if(i == inject_error && !restored) {
                std::cout << "Injected error." << std::endl;
                exit(-1);
            }
        }
    }
    #pragma chk shutdown

    for(i = 0; i < 1024*1024; i++) 
        assert(i == data[i]);

    MPI_Finalize();
}
